pkg/compact/compact_test.go

```diff
@@ -174,3 +179,622 @@ func BenchmarkGatherNoCompactionMarkFilter_Filter(b *testing.B) {
+				keys[0]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+				keys[1]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+				keys[2]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+			},
+		},
+		{
+			testName: "third_test",
+			input: []*metadata.Meta{
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(0, nil),
+						MinTime: 0,
+						MaxTime: int64(2 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"a": "1"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(1, nil),
+						MinTime: int64(2 * time.Hour / time.Millisecond),
+						MaxTime: int64(4 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"b": "2"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(2, nil),
+						MinTime: int64(4 * time.Hour / time.Millisecond),
+						MaxTime: int64(6 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version: 1,
+						Labels:  map[string]string{"b": "2"},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(3, nil),
+						MinTime: int64(6 * time.Hour / time.Millisecond),
+						MaxTime: int64(8 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version:    1,
+						Labels:     map[string]string{"a": "1", "b": "2"},
+						Downsample: metadata.ThanosDownsample{Resolution: 1},
+					},
+				},
+				{
+					BlockMeta: tsdb.BlockMeta{
+						ULID:    ulid.MustNew(4, nil),
+						MinTime: int64(8 * time.Hour / time.Millisecond),
+						MaxTime: int64(10 * time.Hour / time.Millisecond),
+					},
+					Thanos: metadata.Thanos{
+						Version:    1,
+						Labels:     map[string]string{"a": "1", "b": "2"},
+						Downsample: metadata.ThanosDownsample{Resolution: 1},
+					},
+				},
+			},
+			expected: map[string]planResult{
+				keys[0]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+				keys[1]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+				keys[2]: {
+					compactionRuns:   0.0,
+					compactionBlocks: 0.0,
+				},
+			},
+		},
+		{},
+	} {
+		if ok := t.Run(tcase.testName, func(t *testing.T) {
+			blocks := make(map[ulid.ULID]*metadata.Meta, len(tcase.input))
+			for _, meta := range tcase.input {
+				blocks[meta.ULID] = meta
+			}
+			// form groups from the input metadata - do not hardcode groups. hence, grouper.Groups should stay
+			groups, err := grouper.Groups(blocks)
+			if err != nil {
+				level.Warn(logger).Log("msg, unable to form groups")
+			}
+			ps := NewCompactionProgressCalculator(unRegisterer, planner)
+			err = ps.ProgressCalculate(context.Background(), groups)
+			metrics := ps.CompactProgressMetrics
+			testutil.Ok(t, err)
+			for _, key := range keys {
```

Comment:
As we discussed on Slack, it is unnecessary to iterate over all keys and you don't need all 3 groups in all test cases. Use expected map to check the results.